### ISSUE TITLE
JsonSerializationException when deserializing Dictionary<int, HashSet<string>>

### ISSUE THREAD

**Issue by kaufmb** [JamesNK/Newtonsoft.Json#949]
Hi,
I got following  exception on deserializing a property of type `Dictionary<int, HashSet<string>>`:
```
Exception thrown: 'Newtonsoft.Json.JsonSerializationException' in Newtonsoft.Json.dll

Additional information: Could not find type 'System.Collections.Generic.Dictionary`2[[System.Int32, mscorlib],[System.Collections.Generic.HashSet`1[[System.String, mscorlib]], System.Core]]' in assembly 'mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'.
```
When setting TypeNameAssemblyFormat to FormatterAssemblyStyle.Full deserialization works as expected.
In my scenario I cant use this setting.
Is this a bug or are there other ways to solve this?

Kind regards
Bernd

PS: I used following code for reproducing the issue
```
[Test]
public void DeserializeDictionary()
{
    var serializer = new Newtonsoft.Json.JsonSerializer
    {
        TypeNameHandling = TypeNameHandling.All,
        DefaultValueHandling = DefaultValueHandling.Include,
        NullValueHandling = NullValueHandling.Include,
    };

    var dictionary = new Dictionary<int, HashSet<string>>
    {
        { 1, new HashSet<string>(new [] {"test"}) },
    };

    var obtainedJson = Serialize(serializer, dictionary);

    var obtainedDictionary = (Dictionary<int, HashSet<string>>)Deserialize(serializer, obtainedJson);

    Assert.IsNotNull(obtainedDictionary);
}

private static string Serialize(JsonSerializer serializer, object graph)
{
    using (var stringWriter = new StringWriter())
    {
        var jsonWriter = new JsonTextWriter(stringWriter);
        serializer.Serialize(jsonWriter, graph);
        stringWriter.Flush();
        return stringWriter.ToString();
    }
}

private static object Deserialize(JsonSerializer serializer, string input)
{
    using (var stringReader = new StringReader(input))
    {
        var jsonReader = new JsonTextReader(stringReader);
        return serializer.Deserialize(jsonReader);
    }
}
```

**Comment by JamesNK** [comment]
Hi

I'm afraid that is a limitation of Type.GetType. You could try implementing your own serialization binder to fix it, or just use the full assembly name.

**Comment by cremor** [comment]
I agree that this is a limitation of `Type.GetType`. But wouldn't it be a good idea to support this case out of the box with Json.NET?

Since `FormatterAssemblyStyle.Simple` is the default setting, everyone who tries to use a type from the .NET Framework that is not defined in mscorlib will have this problem. Aside from `HashSet` in System.Core.dll there are also a few collection types in System.dll which someone might want to serialize.

Wouldn't it be enough to call `assembly.GetType` for those two assemblies if `Type.GetType` returned null?

**Comment by JamesNK** [comment]
Won't that fail if the collection type in the generic is from another assembly? Or if the collect type is from another assembly?

**Comment by JamesNK** [comment]
Typo: I mean the collection item type

**Comment by cremor** [comment]
You are right, didn't think about that.

So if there is really no API that can resolve such a type string, is it even possible to solve this with an own serialization binder?

**Comment by TylerBrinkley** [comment]
I've made a pull-request to address this issue as well as #926. Please consider reviewing it. Thanks.

**Comment by JamesNK** [comment]
Thanks. I was actually looking at doing the exact same thing last night.

I'll give it a try tonight and get back to you.

**Comment by cremor** [comment]
I've had success with this code, maybe it helps:

    private static void Test()
    {
	    string typeName = "System.Collections.Generic.Dictionary`2[[System.Int32, mscorlib],[System.Collections.Generic.HashSet`1[[System.String, mscorlib]], System.Core]]";
        var type = Type.GetType(typeName, ResolveAssembly, ResolveType);
    }
    
    private static Assembly ResolveAssembly(AssemblyName assemblyName)
	{
        if (assemblyName.Name == assemblyName.FullName)
		{
            return Assembly.LoadWithPartialName(assemblyName.Name);
        }
        return Assembly.Load(assemblyName);
    }
	
    private static Type ResolveType(Assembly assembly, string typeName, bool ignoreCase)
	{
        return assembly != null
            ? assembly.GetType(typeName, false, ignoreCase)
            : Type.GetType(typeName, false, ignoreCase);
    }

**Comment by JamesNK** [comment]
I've merged #982 which fixes this.
